pkg/controller/restore_controller.go

```diff
@@ -506,48 +506,74 @@ func (c *restoreController) fetchFromBackupStorage(backupName string, pluginMana
 		return backupInfo{}, err
 	}
 
-	idx := 1
-	orderedLocations := make([]*api.BackupStorageLocation, len(locations))
-	for i := 0; i < len(locations); i++ {
-		if locations[i].Name == c.defaultBackupLocation {
-			orderedLocations[0] = locations[i]
-		} else {
-			orderedLocations[idx] = locations[i]
-			idx++
-		}
-	}
+	orderedLocations := orderedBackupLocations(locations, c.defaultBackupLocation)
 
 	logContext := c.logger.WithField("backupName", backupName)
 	var info backupInfo
 	for _, location := range orderedLocations {
-		info.objectStore, err = getObjectStoreForLocation(location, pluginManager)
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to get an object store in the location ", location.Name)
-			continue
+		info, err = c.backupInfoForLocation(location, backupName, pluginManager)
+		if err == nil {
+			break
 		}
+		log := logContext.WithField("locationName", location.Name)
+		log.WithError(errors.WithStack(err)).Error("Unable to fetch a backup from object storage")
+	}
 
-		info.backup, err = c.getBackup(info.objectStore, location.Spec.ObjectStorage.Bucket, backupName)
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to find a backup in the location ", location.Name)
-			continue
-		}
+	if info == (backupInfo{}) {
+		return backupInfo{}, errors.New("not able to fetch from backup storage")
+	}
 
-		// ResourceVersion needs to be cleared in order to create the object in the API
-		info.backup.ResourceVersion = ""
-		// Clear out the namespace, in case the backup was made in a different cluster, with a different namespace
-		info.backup.Namespace = ""
+	return info, nil
+}
 
-		info.backup, err = c.backupClient.Backups(c.namespace).Create(info.backup)
-		if err != nil {
-			logContext.WithError(errors.WithStack(err)).Error("Unable to create API object for Backup")
-		} else {
-			// backup found in a location's object store and successfully created in the API
-			info.bucketName = location.Spec.ObjectStorage.Bucket
-			return info, nil
+func orderedBackupLocations(locations []*api.BackupStorageLocation, defaultLocationName string) []*api.BackupStorageLocation {
+	var result []*api.BackupStorageLocation
+
+	for i := range locations {
+		if locations[i].Name == defaultLocationName {
+			// put the default location first
+			result = append(result, locations[i])
+			// append everything before the default
+			result = append(result, locations[:i]...)
+			// append everything after the default
+			result = append(result, locations[i+1:]...)
+
+			return result
 		}
 	}
 
-	return backupInfo{}, errors.New("not able to fetch from backup storage")
+	return locations
+}
+
+func (c *restoreController) backupInfoForLocation(location *api.BackupStorageLocation, backupName string, pluginManager plugin.Manager) (backupInfo, error) {
```

Comment:
I feel much better about splitting this code into a separate function. I wasn't sure if I should have a receiver or just pass the stuff I need as an argument. Any preference?